src/sys/windows/named_pipe.rs

```diff
@@ -0,0 +1,610 @@
+            return Err(would_block());
+        }
+
+        // Now that we've flagged ourselves in the connecting state, issue the
+        // connection attempt. Afterwards interpret the return value and set
+        // internal state accordingly.
+        let res = unsafe {
+            let overlapped = self.inner.connect.as_mut_ptr() as *mut _;
+            self.inner.handle.connect_overlapped(overlapped)
+        };
+
+        match res {
+            // The connection operation finished immediately, so let's schedule
+            // reads/writes and such.
+            Ok(true) => {
+                self.inner.connecting.store(false, SeqCst);
+                Inner::post_register(&self.inner);
+                Ok(())
+            }
+
+            // If the overlapped operation was successful and didn't finish
+            // immediately then we forget a copy of the arc we hold
+            // internally. This ensures that when the completion status comes
+            // in for the I/O operation finishing it'll have a reference
+            // associated with it and our data will still be valid. The
+            // `connect_done` function will "reify" this forgotten pointer to
+            // drop the refcount on the other side.
+            Ok(false) => {
+                mem::forget(self.inner.clone());
+                Err(would_block())
+            }
+
+            // TODO: are we sure no IOCP notification comes in here?
+            Err(e) => {
+                self.inner.connecting.store(false, SeqCst);
+                Err(e)
+            }
+        }
+    }
+
+    /// Takes any internal error that has happened after the last I/O operation
+    /// which hasn't been retrieved yet.
+    ///
+    /// This is particularly useful when detecting failed attempts to `connect`.
+    /// After a completed `connect` flags this pipe as writable then callers
+    /// must invoke this method to determine whether the connection actually
+    /// succeeded. If this function returns `None` then a client is connected,
+    /// otherwise it returns an error of what happened and a client shouldn't be
+    /// connected.
+    pub fn take_error(&self) -> io::Result<Option<io::Error>> {
+        Ok(self.inner.io.lock().unwrap().connect_error.take())
+    }
+
+    /// Disconnects this named pipe from a connected client.
+    ///
+    /// This function will disconnect the pipe from a connected client, if any,
+    /// transitively calling the `DisconnectNamedPipe` function. If the
+    /// disconnection is successful then this object will no longer be readable
+    /// or writable.
+    ///
+    /// After a `disconnect` is issued, then a `connect` may be called again to
+    /// connect to another client.
+    pub fn disconnect(&self) -> io::Result<()> {
+        self.inner.handle.disconnect()
+    }
+
+    /// TODO: dox
+    pub fn read(&mut self, cx: &mut Context<'_>, buf: &mut [u8]) -> Poll<io::Result<usize>> {
```

Comment:
Non of the other types of Mio use the task/`Future` system. So this seems out of place. This should implement the `io::Read` and `io::Write` traits instead?

Reply:
I couldn't think of a different way to implement this w/o any overhaul of mio, adding back the custom readiness queue. That seemed much more overkill than this. Since IOCP is a callback system, we need some way to signal the callback fired. Using `Waker` in std is the easiest way I can think of doing this (it's essentially just an argument-less callback).

Since this is mostly to support Tokio upgrading to Mio 0.7, I would  be OK flagging the type as `doc(hidden)` if you want. Thoughts?

Reply:
This type is primarily here to support the Tokio upgrade, so I could also flag this as `#[doc(hidden)]` for now.

Reply:
We could do something similar to what we do with the `Waker` here: https://github.com/tokio-rs/mio/blob/086251c49dc3ec5dba1483980bc0023a9ea4ee94/src/sys/windows/selector.rs#L491-L505

We would call the callback, but instead of the callback calling `task::Waker.wake`, the code linked above in `SelectorInner::feed_events` would create a new `Event` and add it to the events.

Reply:
I investigated this strategy, there are a number of scenarios in which we need to be able to fire an event w/o having access to the the events list. For example, if `NamedPipe::connect()` is called, a write event needs to be able to fire, but no events list is present. Simply saving off the ready state is not sufficient because something needs to wake up the selector.

Reply:
Ok, I worked around this by using the IOCP handle to queue events.